crates/cdk/src/wallet/subscription/ws.rs

```diff
@@ -0,0 +1,147 @@
+use std::collections::HashMap;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+
+use futures::{SinkExt, StreamExt};
+use tokio::sync::{mpsc, RwLock};
+use tokio_tungstenite::connect_async;
+use tokio_tungstenite::tungstenite::Message;
+use tracing::{debug, error};
+
+use super::WsSubscriptionBody;
+use crate::mint_url::MintUrl;
+use crate::nuts::nut17::ws::{
+    WsMessageOrResponse, WsMethodRequest, WsRequest, WsUnsubscribeRequest,
+};
+use crate::nuts::nut17::Params;
+use crate::pub_sub::SubId;
+
+#[allow(clippy::incompatible_msrv)]
+#[inline]
+pub async fn ws_main(
+    url: MintUrl,
+    subscriptions: Arc<RwLock<HashMap<SubId, WsSubscriptionBody>>>,
+    mut new_subscription_recv: mpsc::Receiver<SubId>,
+    mut on_drop: mpsc::Receiver<SubId>,
+) {
+    let url = url
+        .join_paths(&["v1", "ws"])
+        .as_mut()
+        .map(|url| {
+            if url.scheme() == "https" {
+                url.set_scheme("wss").expect("Could not set scheme");
+            } else {
+                url.set_scheme("ws").expect("Could not set scheme");
+            }
+            url
+        })
+        .expect("Could not join paths")
+        .to_string();
+
+    let mut active_subscriptions = HashMap::<SubId, mpsc::Sender<_>>::new();
+
+    loop {
+        debug!("Connecting to {}", url);
+        let (ws_stream, _) = if let Ok(result) = connect_async(&url).await {
+            result
+        } else {
+            continue;
+        };
+        debug!("Connected to {}", url);
```

Comment:
As mentioned above this creates an infinite loop if the wallet cannot connect to the ws endpoint, I think we should add a timeout.

Reply:
Yeah I think that makes sense. I noticed it because of the above noted issue with a mint that does not signal anything for nut17 and bc of the default we try anyway. In the case of a mint signaling explicitly that it does not support ws we should not try ws at all, if it does signal support we can try and then fall back on failure.